pkg/executor/docker/executor.go

```diff
@@ -95,277 +110,259 @@ func (e *Executor) Run(
+			return fmt.Errorf("execution (%s) already started", request.ExecutionID)
 		} else {
-			return executor.FailResult(fmt.Errorf("unknown storage volume type: %s", input.Volume.Type))
+			// TODO what should we do if an execution has already completed and we try to start it again? Just rerun it?
+			return fmt.Errorf("execution (%s) already completed", request.ExecutionID)
 		}
 	}
 
-	// for this phase of the outputs we ignore the engine because it's just about collecting the
-	// data from the job and keeping it locally
-	// the engine property of the output storage spec is how we will "publish" the output volume
-	// if and when the deal is settled
-	for _, output := range request.Outputs {
-		if output.Name == "" {
-			err = fmt.Errorf("output volume has no name: %+v", output)
-			return executor.FailResult(err)
-		}
-
-		if output.Path == "" {
-			err = fmt.Errorf("output volume has no Location: %+v", output)
-			return executor.FailResult(err)
-		}
-
-		srcd := filepath.Join(request.ResultsDir, output.Name)
-		err = os.Mkdir(srcd, util.OS_ALL_R|util.OS_ALL_X|util.OS_USER_W)
-		if err != nil {
-			return executor.FailResult(err)
-		}
+	jobContainer, err := e.newDockerJobContainer(ctx, &dockerJobContainerParams{
+		ExecutionID:   request.ExecutionID,
+		JobID:         request.JobID,
+		EngineSpec:    request.EngineParams,
+		NetworkConfig: request.Network,
+		Resources:     request.Resources,
+		Inputs:        request.Inputs,
+		Outputs:       request.Outputs,
+		ResultsDir:    request.ResultsDir,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create docker job container: %w", err)
+	}
 
-		log.Ctx(ctx).Trace().Msgf("Output Volume: %+v", output)
+	handler := &executionHandler{
+		client: e.client,
+		logger: log.With().
+			Str("container", jobContainer.ID).
+			Str("execution", request.ExecutionID).
+			Str("job", request.JobID).
+			Logger(),
+		ID:          e.ID,
+		executionID: request.ExecutionID,
+		containerID: jobContainer.ID,
+		resultsDir:  request.ResultsDir,
+		limits:      request.OutputLimits,
+		keepStack:   config.ShouldKeepStack(),
+		waitCh:      make(chan bool),
+		activeCh:    make(chan bool),
+		running:     atomic.NewBool(false),
+	}
 
-		// create a mount so the output data does not need to be copied back to the host
-		mounts = append(mounts, mount.Mount{
+	// register the handler for this executionID
+	e.handlers.Put(request.ExecutionID, handler)
+	// run the container.
+	go handler.run(ctx)
+	return nil
+}
 
-			Type: mount.TypeBind,
-			// this is an output volume so can be written to
-			ReadOnly: false,
+// Wait waits for the completion of a specific execution by its executionID.
+// A channel is returned that will emit the result of the execution.
+// TODO: The current implementation spawns a new Goroutine for each Wait call,
+// which could become resource-intensive with a large number of concurrent Wait calls.
```

Comment:
I don't think this is much of a problem, will we really create many of this per execution?  Even if for some reason we end up with dozens of goroutines it doesn't seem excessive. We have a broadcaster in pkg/util/generic/broadcaster.go which hands out channels, but is super-cautious to avoid writing to closed channels and as a result is probably not as performant as it could be. 

Is there a case that would generate many goroutines that I'm missing?

Reply:
I can't think of any, was leaving it more as a hint for future implementers. I have removed the TODO and changed the comment to be more specific to the function of the method.